### PR TITLE
Simultaneous deriv support

### PR DESCRIPTION
**Pull request by naylor-b**
- also fixed a bug that prevented numpy.int32 and numpy.int64 from being used as indices.
- simul derivs work only in fwd mode currently.

### CODE REVIEW

File: openmdao/drivers/tests/test_pyoptsparse_driver.py [hunk 1]
```diff
@@ -1020,7 +1020,7 @@ def setup(self):
 
                 self.declare_partials('*', '*')
 
-                self.fail_deriv = [2, 4]
+                self.fail_deriv = [3, 5]
```
Comment (Kenneth-T-Moore) [inline openmdao/drivers/tests/test_pyoptsparse_driver.py]: Why did you change which derivs fail on this test?

File: openmdao/docs/theory_manual/simul_derivs.rst [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                # design variable
+                'indeps.x': {
+                    0: ([0, 1, 2, 3, 4], [0, 2, 4, 6, 8]),
+                    1: ([0, 1, 2, 3, 4], [1, 3, 5, 7, 9])
+                }
+            }
+        }
+    )
+
+    prob.driver.set_simul_coloring(color_info)
```
Comment (hwangjt) [inline openmdao/docs/theory_manual/simul_derivs.rst]: Perhaps it would be good to add a feature doc in addition to this theory doc or turn this into a feature doc, since this is a user-facing method. The function signature can be included in the feature doc then

Comment (hwangjt) [reply openmdao/docs/theory_manual/simul_derivs.rst]: Are these names promoted or absolute names?

File: openmdao/drivers/tests/test_pyoptsparse_driver.py (hunk 1 above)
Comment (naylor-b) [reply openmdao/drivers/tests/test_pyoptsparse_driver.py]: I don't remember why I had to do that, but I just reverted the change and the test passes, so I'll get rid of that change.

File: openmdao/docs/theory_manual/simul_derivs.rst (hunk 2 above)
Comment (naylor-b) [reply openmdao/docs/theory_manual/simul_derivs.rst]: I will move this to a feature doc and add the function signature.  The names in the coloring data structure are currently absolute.  I could add code to do the mapping from promoted to absolute if that's preferred.

Comment (hwangjt) [reply openmdao/docs/theory_manual/simul_derivs.rst]: Great. Yes, I think promoted would be easier for users

File: openmdao/core/driver.py [hunk 3]
```diff
@@ -418,14 +439,15 @@ def set_design_var(self, name, value):
         desvar = self._problem.model._outputs._views_flat[name]
         desvar[indices] = value
 
-        # Scale design variable values
-        scaler = meta['scaler']
-        if scaler is not None:
-            desvar[indices] *= 1.0 / scaler
+        if self._has_scaling:
```
[the remark this answers is not in the document]
Comment (naylor-b) [reply openmdao/core/driver.py]: yeah

File: openmdao/core/driver.py [hunk 4]
```diff
@@ -741,3 +752,30 @@ def _get_name(self):
...
+    def set_simul_coloring(self, simul_info):
+        """
+        Set the coloring for simultaneous derivatives.
+
+        Parameters
+        ----------
+        simul_info : tuple
+            Information about simultaneous coloring for design vars and responses.
+        """
+        if self.supports['simultaneous_derivatives']:
```
[the remark this answers is not in the document]
Comment (naylor-b) [reply openmdao/core/driver.py]: This was done as a quick and dirty way to make sure someone doesn't try using simul derivs with a driver that doesn't split up linear and nonlinear constraints like pyoptsparse does. We had talked about adding this to the scipy optimizer at some point.

File: openmdao/core/system.py [hunk 5]
```diff
@@ -2315,7 +2339,8 @@ def add_objective(self, name, ref=None, ref0=None, index=None,
         self.add_response(name, type_='obj', scaler=scaler, adder=adder,
                           ref=ref, ref0=ref0, index=index,
                           parallel_deriv_color=parallel_deriv_color,
-                          vectorize_derivs=vectorize_derivs)
+                          vectorize_derivs=vectorize_derivs,
+                          simul_coloring=simul_coloring, simul_map=simul_map)
```
[the remark this answers is not in the document]
Comment (naylor-b) [reply openmdao/core/system.py]: I agree that simul_map isn't a very descriptive name, but I don't think simul_io_map makes things any clearer.
